File: mobile/lib/src/accounting.dart
import 'models.dart';

class AccountPosition {
  AccountPosition(Workspace workspace) : metrics = workspace.metrics;
  final JsonMap metrics;
  bool get ready => metrics.containsKey('ledgerBalance');
  int get balance => readInt(metrics, 'ledgerBalance');
  int get due => balance > 0 ? balance : 0;
  int get credit => balance < 0 ? -balance : 0;
  int get charges => readInt(metrics, 'totalCharges');
  int get adjustments => readInt(metrics, 'totalAdjustments');
  int get received => readInt(metrics, 'totalReceived');
  int get pending => readInt(metrics, 'pendingPaymentAmount');
}

File: frontend/src/platform/accounting.ts
import type { Workspace } from "./types";

export const accountingInvalidatedEvent = "cc:accounting-invalidated";
export const financialScopes = new Set(["manufacturing", "shipping", "payments", "ledger"]);

// Every customer balance surface reads the same authoritative server summary.
export function accountPosition(workspace: Workspace) {
  const m = workspace.metrics || {};
  const ready = Object.prototype.hasOwnProperty.call(m, "ledgerBalance");
  const balance = ready ? Number(m.ledgerBalance) : 0;
  return {
    ready, balance,
    due: Math.max(0, balance), credit: Math.max(0, -balance),
    charges: Number(m.totalCharges || 0),
    adjustments: Number(m.totalAdjustments || 0),
    received: Number(m.totalReceived || 0),
    pending: Number(m.pendingPaymentAmount || 0)
  };
}
